rootfs/ingest/python/qserv/ingest.py

```diff
@@ -62,175 +55,181 @@
+
+            contributions = build_contributions(contributions_locked,
+                                                self.repl_client,
+                                                self.contribution_metadata.load_balanced_url)
+            ingest_success = self._ingest_all_contributions(transaction_id, contributions)
         except Exception as e:
             _LOG.critical('Ingest failed during transaction: %s, %s', transaction_id, e)
             ingest_success = False
             # Stop process when any transaction abort
             raise(e)
         finally:
             if transaction_id:
-                self.repl_client.close_transaction(self.chunk_meta.database,
-                                       transaction_id,
-                                       ingest_success)
-                self.queue_manager.release_locked_chunkfiles(ingest_success)
+                self.repl_client.close_transaction(self.contribution_metadata.database,
+                                                   transaction_id,
+                                                   ingest_success)
+                self.queue_manager.release_locked_contributions(ingest_success)
 
         return True
 
-    def _ingest_parallel_contributions(self, transaction_id: int, chunks_locked):
-        args = self._compute_args(chunks_locked, transaction_id)
-        with ThreadPool() as pool:
-            for chunk, startedAt, endedAt in pool.imap_unordered(_ingest_contribution, args):
-                logging.debug('Chunk %s ingest started at %s ended at %s' % (chunk, startedAt, endedAt))
+    def _ingest_all_contributions(self, transaction_id: int, contributions: list) -> bool:
+        """Ingest all contribution for a given transaction
+           Throw exception if ingest fail
+           This method always returns True, or raise an exception
+
+        Args:
+            transaction_id (int): id of the transaction
+            contributions (list): list of contribution to ingest
+
+        Returns:
+            bool: True if ingest has ran successfully
+        """
+        finished = False
+        contrib_ingested_count = len(contributions)
+        while not finished:
+            idx_contribs_in_progress = []
+            for i, c in enumerate(contributions):
+                # Ingest to start
+                if c.request_id is None:
+                    # Ingest to start
+                    startedAt = time.strftime("%H:%M:%S", time.localtime())
+                    _LOG.debug('Contribution %s ingest started at %s', c, startedAt)
+                    c.start_async(transaction_id)
+                    idx_contribs_in_progress.append(i)
+                else:
+                    startedAt = time.strftime("%H:%M:%S", time.localtime())
+                    _LOG.debug('Contribution %s ingest monitored at %s', c, startedAt)
+                    if c.monitor():
```

Comment:
```
    def monitor(self) -> bool:
        """Monitor an asynchronous ingest query for a chunk contribution
           States of an ingest query are listed here:
           https://confluence.lsstcorp.org/display/DM/Ingest%3A+9.5.3.+Asynchronous+Protocol
           Propagate an exception if the query fails after a fixed number of attempts

        Raises:
            Exception: in case of error during contribution ingest

        Returns:
            bool: True if contribution has been successfully ingested
                  False if contribution is being ingested
        """
```